[Fix #4368] Extend flexibility of `Performance/HashEachMethod cop

gohdaniel15 opened the pull request:
This PR extends the node matching capability of the `Performance/HashEachMethod` cop. Prior to this, the node matcher would only work on `Hash`. This has been extended to match any node. 

This cop was then run on the codebase, and the fixes were implemented in a70086b. 
Please see comments that I have left on the individual files. 

-----------------

Before submitting the PR make sure the following are checked:

* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Used the same coding conventions as the rest of the project.
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [X] All tests(`rake spec`) are passing.
* [X] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

gohdaniel15 commented on lib/rubocop/cop/lint/percent_string_array.rb:
```diff
@@ -55,7 +55,7 @@ def contains_quotes_or_commas?(node)
+            node.each_child_node do |value|
               range = value.loc.expression
 
               match = range.source.match(TRAILING_QUOTE)
```
```Ruby
# Original lib/rubocop/cop/lint/percent_string_array.rb
def autocorrect(node)
  lambda do |corrector|
    node.values.each do |value|
      ...
```

Calling `#values` on `ArrayNode` simply calls `#each_child_node.to_a`. It is pointless to call `#to_a` if you are just going to call `#each` on it after, therefore, just use `#each_child_node`

```Ruby
# Modified lib/rubocop/cop/lint/percent_string_array.rb
def autocorrect(node)
  lambda do |corrector|
    node.each_child_node do |value|
      ...
```

Drenmi replied:
This means we're losing valuable information, and we're back to just fiddling with `children` of anonymous nodes, which is part of the problem node extensions solve. 

 The `#values` method should probably be `#elements`. (I can't remember if it's used for polymorphism, but if we change it, we'll see which tests break.)

Also feel free to change the `node` parameter to `array_node`. 🙂

gohdaniel15 replied:
> (I can't remember if it's used for polymorphism, but if we change it, we'll see which tests break.)
You are right. Rails/EnumUniqueness incorporates this
```
class EnumUniqueness < Cop
...
  def on_send(node)
    enum_declaration(node) do |name, 
      items = args.values
...
```

gohdaniel15 commented on lib/rubocop/config_loader.rb:
```diff
@@ -60,7 +60,7 @@ def create_config(hash, path)
+        hash.clone.each_key do |k|
           q = Cop::Cop.qualified_cop_name(k, path)
           next if q == k
 
```
```Ruby
# Original lib/rubocop/config_loader.rb
def add_missing_namespaces(path, hash)
  hash.keys.each do |k|
    q = Cop::Cop.qualified_cop_name(k, path)
    next if q == k

    hash[q] = hash.delete(k)
  end
end
```

The modified Lint/HashEachMethod cop will catch `hash.keys.each` and correct it to `hash.each_key`. This will work- until it runs `hash[q] = hash.delete(k)`, which will spout `RuntimeError: can't add a new key into hash during iteration`. This is because `lazy iterations` have been implemented in Ruby 1.9, forbidding new keys to be added to the hash that is being iterated over. 

One solution that I have looked up is to clone it before iterating over it. Not a big fan of this though. Any suggestions?

```Ruby
# Modified lib/rubocop/config_loader.rb
def add_missing_namespaces(path, hash)
  hash.clone.each_key do |k|
    q = Cop::Cop.qualified_cop_name(k, path)
    next if q == k

    hash[q] = hash.delete(k)
  end
end
```

Drenmi replied:
Mark as false positive by wrapping in `# rubocop:disable Performance/HashEachMethod` and `# enable Performance/HashEachMethod`. 🙂

Also, feel free to change the single character variables to meaningful names.

Drenmi replied:
(Anecdotally: I think ActiveSupport added a monkey patch for this. `Hash#transform_keys`.)

gohdaniel15 replied:
👍  thank you

Drenmi commented on lib/rubocop/cop/performance/hash_each_methods.rb:
```diff
@@ -55,11 +67,11 @@ def autocorrect(node)
-            if receiver.method?(:hash)
-              correct_hash(node, corrector)
-            else
+            if (first_method == :keys) || (first_method == :values)
```
Should be no need for parentheses here. (False negative of `Style/RedundantParentheses`? 🤔 )

Drenmi replied:
Should be no need for parentheses here. (False negative in `Style/RedundantParentheses`? 🤔 )

gohdaniel15 replied:
Hmm yes, perhaps. Let me dig a bit deeper into this.

Drenmi replied:
Don't get side tracked though ... 😉

Drenmi commented on CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@
 * [#4568](https://github.com/bbatsov/rubocop/pull/4568): Fix autocorrection for `Style/TrailingUnderscoreVariable`. ([@smakagon][])
 * [#4586](https://github.com/bbatsov/rubocop/pull/4586): Add new `Performance/UnfreezeString` cop. ([@pocke][])
 * [#2976](https://github.com/bbatsov/rubocop/issues/2976): Add `Whitelist` configuration option to `Style/NestedParenthesizedCalls` cop. ([@drenmi][])
+* [#4368](https://github.com/bbatsov/rubocop/issues/4368): Extend flexibility of `Performance/HashEachMethod` cop. ([@gohdaniel15][])
```
Try to outline what was changed in the description. E.g.: "Make `Performance/HashEachMethod` inspect all send nodes."

Drenmi commented:
Great stuff! ❤️ 

Can you clean up the test cases a bit? It seems that there are these contexts:

- Signature:
  - `each do |key, _value|`
  - `keys.each`
- Receiver:
  - hash literal (`{}.each ...`)
  - another send (`foo.each ...`)
  - implicit (`each ...`)

The descriptions have become a bit confounded, e.g. `Hash#each` for send nodes named `hash`, etc. And since we're not restricting it to `hash` anymore, let's use a metasyntactic variable (like `foo`) to make it clear in the tests that it doesn't matter. 🙂

gohdaniel15 commented:
@Drenmi Made the changes 🙂 Apologies it took so long, have been swamped with work as of late. Do have a look when you have the time.

Drenmi commented on lib/rubocop/cop/lint/percent_string_array.rb:
```diff
@@ -53,6 +53,7 @@ def contains_quotes_or_commas?(node)
+        # rubocop:disable Performance/HashEachMethods
         def autocorrect(node)
           lambda do |corrector|
             node.values.each do |value|
```
@drenmi: Note to self. Make node extensions that respond to `#values` also respond to `#each_value`.

Drenmi commented on lib/rubocop/cop/performance/hash_each_methods.rb:
```diff
@@ -55,17 +67,32 @@ def autocorrect(node)
-            if receiver.method?(:hash)
-              correct_hash(node, corrector)
-            else
+            if first_method.eql?(:keys) || first_method.eql?(:values)
```
Not sure why hash equality (`#eql?`) is used to compare symbols? 🙂 

```ruby
%i[keys values].include?(first_method)
```

or

```ruby
first_method == :keys || first_method == :values
```

Same applies to all uses of `#eql?`.

Drenmi commented on lib/rubocop/cop/performance/hash_each_methods.rb:
```diff
@@ -55,17 +67,32 @@ def autocorrect(node)
-              correct_hash(node, corrector)
-            else
+            if first_method.eql?(:keys) || first_method.eql?(:values)
+              return correct_implicit(node, corrector) if receiver
```
What is the predicate `receiver.child_nodes.first.nil?`?

Maybe there is a node extension method that can be used to clarify this? Or it can be extracted into a named method? 🙂

gohdaniel15 replied:
When its a plain_each match, then we check if receiver is nil. 
When its a key_value match, I have to check if the main receiver of (key || value) is nil. 

I have changed it to `receiver.receiver.nil?`. Not sure if there is a better node extension method that can be used in place here?

Drenmi commented on lib/rubocop/cop/performance/hash_each_methods.rb:
```diff
@@ -55,17 +67,32 @@ def autocorrect(node)
               new_source = caller.source + ".each_#{first_method[0..-2]}"
               corrector.replace(node.loc.expression, new_source)
+            else
+              return correct_implicit(node, corrector) if receiver.nil?
```
Maybe this could be changed to an `elsif` + `else`?

```ruby
if ...
  ...
elsif !receiver
  correct_implicit(node, corrector)
else
  correct_plain_each(node, corrector)
end
```

gohdaniel15 replied:
Hmm, as a broad overview, what I am trying to do is:

```Ruby
case receiver
when node_matches_key_value?
  if receiver.implicit?
    autocorrect_implicit
  else
    correct_key_value
  end
when node_matches_plain_each?
  if receiver.implicit?
    autocorrect_implicit
  else
    correct_plain_each
  end
end
```

WDYT of this?

gohdaniel15 commented:
@Drenmi I've made some improvements to the code, and addressed the issues you brought up in the comments. However, I am still unsatisfied with the `#autocorrect` method. I think I might have improved its legibility (or perhaps I've just worked with it too much that its only more legible to me). 

Let me know your thoughts 😄

bbatsov commented on spec/rubocop/cop/performance/hash_each_methods_spec.rb:
```diff
@@ -3,67 +3,467 @@
+      end
+    end
   end
+
```
This should not be part of the commit.

bbatsov commented:
Fix the first commit and rebase on top of the current `master`, please.

gohdaniel15 commented:
@bbatsov My apologies. This is now fixed and rebased 🙂
